File: [0005][Medium][Longest_Palindromic_Substring][DP,string]/Longest_Palindromic_Substring_2.java
// leetcode time     cost : 31 ms
// leetcode memory   cost : 38.1 MB
// Time  Complexity: O(n**2)
// Space Complexity: O(n)
// solution 2, 2 pointer，中心扩散法，要注意奇数及偶数长度回文
public class Solution {
    private int lo, maxLen;

    public String longestPalindrome(String s) {
        int len = s.length();
        if (len < 2)
            return s;
        
        for (int i = 0; i < len-1; i++) {
            extendPalindrome(s, i, i);  //assume odd length, try to extend Palindrome as possible
            extendPalindrome(s, i, i+1); //assume even length.
        }
        return s.substring(this.lo, this.lo + this.maxLen);
    }

    private void extendPalindrome(String s, int j, int k) {
        while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
            j--;
            k++;
        }
        if (this.maxLen < k - j - 1) {
            this.lo = j + 1;
            this.maxLen = k - j - 1;
        }
    }
}

File: [0005][Medium][Longest_Palindromic_Substring][DP,string]/Longest_Palindromic_Substring_2.py
# -*- coding: utf-8 -*-  
# leetcode time     cost : 1008 ms
# leetcode memory   cost : 13.5 MB
# solution 2，中心扩散法，要注意奇数及偶数长度回文
class Solution:
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        res = ""
        for i in range(len(s)):
            res = max(self.helper(s,i,i), self.helper(s,i,i+1), res, key=len)
        return res
       
    def helper(self,s,l,r):        
        while 0<=l and r < len(s) and s[l]==s[r]:
                l-=1; r+=1
        return s[l+1:r]

def main():
    inputX,expectRes = "babad","aba" # ans is "aba" or "bab"
    obj = Solution()
    result = obj.longestPalindrome(inputX)
    try:
        assert result == expectRes
        print("passed, result is follow expect:",result)
    except AssertionError as aError:
        print('failed, result >> ', result,"<< is wrong, ","expect is : "+ expectRes)
    
if __name__ =='__main__':
    main() 